README.md
# args_simple

[![pub package](https://img.shields.io/pub/v/args_simple.svg?logo=dart&logoColor=00b9fc)](https://pub.dev/packages/args_simple)
[![Null Safety](https://img.shields.io/badge/null-safety-brightgreen)](https://dart.dev/null-safety)
[![Codecov](https://img.shields.io/codecov/c/github/gmpassos/args_simple)](https://app.codecov.io/gh/gmpassos/args_simple)
[![Dart CI](https://github.com/gmpassos/args_simple/actions/workflows/dart.yml/badge.svg?branch=master)](https://github.com/gmpassos/args_simple/actions/workflows/dart.yml)
[![GitHub Tag](https://img.shields.io/github/v/tag/gmpassos/args_simple?logo=git&logoColor=white)](https://github.com/gmpassos/args_simple/releases)
[![New Commits](https://img.shields.io/github/commits-since/gmpassos/args_simple/latest?logo=git&logoColor=white)](https://github.com/gmpassos/args_simple/network)
[![Last Commits](https://img.shields.io/github/last-commit/gmpassos/args_simple?logo=git&logoColor=white)](https://github.com/gmpassos/args_simple/commits/master)
[![Pull Requests](https://img.shields.io/github/issues-pr/gmpassos/args_simple?logo=github&logoColor=white)](https://github.com/gmpassos/args_simple/pulls)
[![Code size](https://img.shields.io/github/languages/code-size/gmpassos/args_simple?logo=github&logoColor=white)](https://github.com/gmpassos/args_simple)
[![License](https://img.shields.io/github/license/gmpassos/args_simple?logo=open-source-initiative&logoColor=green)](https://github.com/gmpassos/args_simple/blob/master/LICENSE)

A simple argument parser and handler, integrated with JSON and dart:io. 

## API Documentation

See the [API Documentation][api_doc] for a full list of functions, classes and extension.

[api_doc]: https://pub.dev/documentation/args_simple/latest/

## Usage

```dart
import 'dart:io';

import 'package:args_simple/args_simple_io.dart';

/// Examples of valid values for the parameter `args`:
///
/// - `/path/config.json --port 81 -verbose`
/// - `/path/config.json -verbose --sys-config /my/sys-config-dir`
/// - `/path/config.json root`
void main(List<String> argsOrig) {
  var args = ArgsSimple.parse(argsOrig);

  if (args.isEmpty) {
    print(
            'USAGE: [%configFile.json] [%user] [--port] [--sys-config %systemConfig] [-verbose]');
    exit(0);
  }

  // Argument #0 is a JSON `File`:
  var config = args.argumentMatches(0, RegExp(r'.json$'))
          ? args.argumentAsFileContentJSON(0)!
          : {};

  // Argument #1 is an optional `String`, with `guest` as default:
  var user = args.argumentAsString(1, 'guest');

  // Option `sys-config` is a `Directory`, with `/default/sys-config-dir` as default:
  var systemConfigDir = args.optionAsDirectory(
          'sys-config', Directory('/default/sys-config-dir'))!;

  // Option `--port` is an `int`, with 8080 as default:
  var port = args.optionAsInt('port', 8080);

  // Check for flag `-verbose`:
  var verbose = args.flag('verbose');

  if (verbose) {
    print('-- Config: $config');
    print('-- User: $user');
    print('-- System-Config Dir: $systemConfigDir');
    print('-- Port: $port');
    print('-- Verbose: $verbose');
    print(args);
  }
}
```

## Source

The official source code is [hosted @ GitHub][github_args_simple]:

- https://github.com/gmpassos/args_simple

[github_args_simple]: https://github.com/gmpassos/args_simple

# Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

# Contribution

Any help from the open-source community is always welcome and needed:
- Found an issue?
    - Please fill a bug report with details.
- Wish a feature?
    - Open a feature request with use cases.
- Are you using and liking the project?
    - Promote the project: create an article, do a post or make a donation.
- Are you a developer?
    - Fix a bug and send a pull request.
    - Implement a new feature, like other training algorithms and activation functions.
    - Improve the Unit Tests.
- Have you already helped in any way?
    - **Many thanks from me, the contributors and everybody that uses this project!**

*If you donate 1 hour of your time, you can contribute a lot,
because others will do the same, just be part and start with your 1 hour.*

[tracker]: https://github.com/gmpassos/args_simple/issues

# Author

Graciliano M. Passos: [gmpassos@GitHub][github].

[github]: https://github.com/gmpassos

## License

[Apache License - Version 2.0][apache_license]

[apache_license]: https://www.apache.org/licenses/LICENSE-2.0.txt


CHANGELOG.md
## 1.1.0

- `ArgsSimple`:
  - `parse`: allow multiple values for the same option.
  - Added `flagOr`.

- sdk: '>=3.0.0 <4.0.0'
- collection: ^1.18.0
- yaml: ^3.1.2
- lints: ^2.1.1
- test: ^1.24.9
- dependency_validator: ^3.2.2
- coverage: ^1.7.1

## 1.0.0

- Initial version.


test/args_simple_test.dart
import 'dart:io';

import 'package:args_simple/args_simple_io.dart';
import 'package:test/test.dart';

void main() {
  group('ArgsSimple', () {
    test('basic', () {
      var args = ArgsSimple.parse(['a', 'b', '123', '-123']);
      expect(args.arguments, equals(['a', 'b', '123', '-123']));
      expect(args.options, isEmpty);
      expect(args.flags, isEmpty);
      expect(args.properties, isEmpty);

      expect(args.toList(), equals(['a', 'b', '123', '-123']));
    });

    test('with options', () {
      var args = ArgsSimple.parse(['a', '--b', '123', '456', '--x']);
      expect(args.arguments, equals(['a', '456']));
      expect(args.options, equals({'b': '123', 'x': ''}));
      expect(args.flags, isEmpty);
      expect(args.properties, isEmpty);
    });

    test('with flags', () {
      var args = ArgsSimple.parse(['a', '--b', '123', '456', '-x', '-y']);
      expect(args.arguments, equals(['a', '456']));
      expect(args.options, equals({'b': '123'}));
      expect(args.flags, equals({'x', 'y'}));

      expect(args.optionAsInt('b'), equals(123));
      expect(args.optionAsInt('B'), equals(123));
    });

    test('with properties', () {
      var args = ArgsSimple.parse(
          ['a', '--B', '123', '456', '-X', '-y', '-PZzz=WWW', '-DeA']);
      expect(args.arguments, equals(['a', '456']));
      expect(args.options, equals({'b': '123'}));
      expect(args.flags, equals({'x', 'y'}));
      expect(args.properties, equals({'zzz': 'WWW', 'ea': ''}));

      expect(args.toList(),
          equals(['a', '456', '-x', '-y', '--b', '123', '-Pzzz=WWW', '-Pea=']));

      expect(
          args.toJson(),
          equals({
            'arguments': ['a', '456'],
            'flags': {'x', 'y'},
            'options': {'b': '123'},
            'properties': {'zzz': 'WWW', 'ea': ''}
          }));
    });

    test('with File', () {
      var args = ArgsSimple.parse(['a', '--file', 'pubspec.yaml', '456']);
      expect(args.arguments, equals(['a', '456']));

      expect(args.optionAsFile('file')?.path, equals('pubspec.yaml'));
      expect(args.optionAsDirectory('file')?.path, equals('pubspec.yaml'));

      expect(args.toList(), equals(['a', '456', '--file', 'pubspec.yaml']));
    });

    test('loadEnvironment', () {
      var args = ArgsSimple.parse(['a', '--file', 'pubspec.yaml', '456']);
      expect(args.arguments, equals(['a', '456']));

      var size = args.properties.length;

      var loadedKeys = args.loadEnvironment();

      expect(loadedKeys.length, greaterThan(1));

      expect(args.properties.length, greaterThan(size));
    });

    test('isEmpty', () {
      expect(ArgsSimple().isEmpty, isTrue);
      expect(ArgsSimple.parse([]).isEmpty, isTrue);
    });

    test('parse 1', () {
      var args = ArgsSimple.parse('--port 81 -verbose'.split(RegExp(r'\s+')));
      print(args);

      expect(args.isNotEmpty, isTrue);

      var config = args.argumentMatches(0, RegExp(r'.json$'))
          ? args.argumentAsFileContentJSON(0)!
          : {};

      expect(config, equals({}));

      expect(args.argumentAsString(1, 'guest'), 'guest');

      expect(
          args
              .propertyAsDirectory(
                  'sys-config', Directory('/default/sys-config-dir'))!
              .path,
          endsWith('sys-config-dir'));

      expect(args.optionAsInt('port', 8080), equals(81));

      expect(args.flag('verbose'), isTrue);
    });

    test('parse 2', () {
      var args = ArgsSimple.parse('file.json root'.split(RegExp(r'\s+')));
      print(args);

      expect(args.isNotEmpty, isTrue);

      var configFile = args.argumentMatches(0, RegExp(r'.json$'))
          ? args.argumentAsFile(0)!
          : null;

      expect(configFile!.path, endsWith('file.json'));

      expect(args.argumentAsString(1, 'guest'), 'root');

      expect(
          args
              .propertyAsDirectory(
                  'sys-config', Directory('/default/sys-config-dir'))!
              .path,
          endsWith('sys-config-dir'));

      expect(args.optionAsInt('port', 8080), equals(8080));

      expect(args.flag('verbose'), isFalse);
    });

    test('parse 2', () {
      var args =
          ArgsSimple.parse('file.json -Psys-config=/etc'.split(RegExp(r'\s+')));
      print(args);

      expect(args.isNotEmpty, isTrue);

      var configFile = args.argumentMatches(0, RegExp(r'.json$'))
          ? args.argumentAsFile(0)!
          : null;

      expect(configFile!.path, endsWith('file.json'));

      expect(args.argumentAsString(1, 'guest'), 'guest');

      expect(
          args
              .propertyAsDirectory(
                  'sys-config', Directory('/default/sys-config-dir'))!
              .path,
          endsWith('etc'));

      expect(args.optionAsInt('port', 8080), equals(8080));

      expect(args.flag('verbose'), isFalse);
    });
  });
}


lib/src/type_parser.dart
import 'package:collection/collection.dart';

typedef TypeElementParser<T> = T? Function(Object? o);

typedef TypeParserResolver = TypeElementParser? Function(
    {Type? t, Object? obj, Type? type});

/// Lenient parsers for basic Dart types.
class TypeParser {
  static final Set<TypeParserResolver> _parsersResolvers =
      <TypeParserResolver>{};

  static void registerParserResolver(TypeParserResolver parserResolver) =>
      _parsersResolvers.add(parserResolver);

  static bool unregisterParserResolver(TypeParserResolver parserResolver) =>
      _parsersResolvers.remove(parserResolver);

  /// Returns the parser for the desired type, defined by [T], [obj] or [type].
  static TypeElementParser? parserFor<T>({Object? obj, Type? type}) {
    if (obj != null) {
      if (obj is String) {
        return parseString;
      } else if (obj is Map) {
        return parseMap;
      } else if (obj is Set) {
        return parseSet;
      } else if (obj is List || obj is Iterable) {
        return parseList;
      } else if (obj is int) {
        return parseInt;
      } else if (obj is double) {
        return parseDouble;
      } else if (obj is num) {
        return parseNum;
      } else if (obj is bool) {
        return parseBool;
      } else if (obj is BigInt) {
        return parseBigInt;
      } else if (obj is DateTime) {
        return parseDateTime;
      }
    }

    type ??= obj?.runtimeType ?? T;

    if (type == String) {
      return parseString;
    } else if (type == Map) {
      return parseMap;
    } else if (type == Set) {
      return parseSet;
    } else if (type == List) {
      return parseList;
    } else if (type == int) {
      return parseInt;
    } else if (type == double) {
      return parseDouble;
    } else if (type == num) {
      return parseNum;
    } else if (type == bool) {
      return parseBool;
    } else if (type == BigInt) {
      return parseBigInt;
    } else if (type == DateTime) {
      return parseDateTime;
    }

    for (var r in _parsersResolvers) {
      var parser = r(t: T, type: type, obj: obj);
      if (parser != null) return parser;
    }

    return null;
  }

  static final RegExp _regexpListDelimiter = RegExp(r'\s*[,;]\s*');

  /// Tries to parse a [List].
  /// - Returns [def] if [value] is `null` or an empty [String].
  /// - [elementParser] is an optional parser for the elements of the parsed [List].
  static List<T>? parseList<T>(Object? value,
      {List<T>? def, TypeElementParser<T>? elementParser}) {
    if (value == null) return def;

    var l = _parseListImpl(value);
    if (l == null) return def;

    if (elementParser != null) {
      l = l.map(elementParser).toList();
    }

    if (l is List<T>) {
      return l;
    } else if (elementParser != null) {
      l = l.whereType<T>().toList();
    } else {
      var parser = parserFor<T>();
      if (parser != null) {
        l = l.map(parser).toList();
      }
    }

    if (l is List<T>) {
      return l;
    } else {
      return l.whereType<T>().toList();
    }
  }

  static List? _parseListImpl<T>(Object value) {
    if (value is List) {
      return value;
    } else if (value is Iterable) {
      return value.toList();
    } else {
      var s = '$value'.trim();
      if (s.isEmpty) return null;
      var l = s.split(_regexpListDelimiter);
      return l;
    }
  }

  /// Tries to parse a [Set].
  ///
  /// See [parseList].
  static Set<T>? parseSet<T>(Object? value,
      {Set<T>? def, TypeElementParser<T>? elementParser}) {
    var l = parseList<T>(value, elementParser: elementParser);
    return l?.toSet() ?? def;
  }

  static final RegExp _regexpPairDelimiter = RegExp(r'\s*[;&]\s*');
  static final RegExp _regexpKeyValueDelimiter = RegExp(r'\s*[:=]\s*');

  /// Tries to parse a [Map].
  /// - Returns [def] if [value] is `null` or an empty [String].
  static Map<K, V>? parseMap<K, V>(Object? value,
      {Map<K, V>? def,
      TypeElementParser<K>? keyParser,
      TypeElementParser<V>? valueParser}) {
    if (value == null) return def;

    if (value is Map<K, V>) {
      return value;
    }

    keyParser ??= parserFor<K>() as TypeElementParser<K>?;
    keyParser ??= (k) => k as K;

    valueParser ??= parserFor<V>() as TypeElementParser<V>?;
    valueParser ??= (v) => v as V;

    if (value is Map) {
      return value
          .map((k, v) => MapEntry(keyParser!(k) as K, valueParser!(v) as V));
    } else if (value is Iterable) {
      return Map.fromEntries(value
          .map((e) => parseMapEntry<K, V>(e,
              keyParser: keyParser, valueParser: valueParser))
          .whereNotNull());
    } else {
      var s = '$value'.trim();
      if (s.isEmpty) return def;

      var pairs = s.split(_regexpPairDelimiter);
      return Map.fromEntries(pairs
          .map((e) => parseMapEntry<K, V>(e,
              keyParser: keyParser, valueParser: valueParser))
          .whereNotNull());
    }
  }

  /// Tries to parse a [MapEntry].
  /// - Returns [def] if [value] is `null` or an empty [String].
  static MapEntry<K, V>? parseMapEntry<K, V>(Object? value,
      {MapEntry<K, V>? def,
      TypeElementParser<K>? keyParser,
      TypeElementParser<V>? valueParser}) {
    if (value == null) return def;

    if (value is MapEntry<K, V>) {
      return value;
    }

    keyParser ??= parserFor<K>() as TypeElementParser<K>?;
    keyParser ??= (k) => k as K;

    valueParser ??= parserFor<V>() as TypeElementParser<V>?;
    valueParser ??= (v) => v as V;

    if (value is MapEntry) {
      return MapEntry(keyParser(value.key) as K, valueParser(value.value) as V);
    } else if (value is Iterable) {
      var k = value.elementAt(0);
      var v = value.elementAt(1);
      return MapEntry(keyParser(k) as K, valueParser(v) as V);
    } else {
      var s = '$value'.trim();
      if (s.isEmpty) return def;

      var idx = s.indexOf(_regexpKeyValueDelimiter);
      if (idx >= 0) {
        var k = s.substring(0, idx);
        var v = s.substring(idx + 1);
        return MapEntry(keyParser(k) as K, valueParser(v) as V);
      }
      return MapEntry(keyParser(s) as K, null as V);
    }
  }

  /// Tries to parse a [String].
  /// - Returns [def] if [value] is `null`.
  static String? parseString(Object? value, [String? def]) {
    if (value == null) return def;

    if (value is String) {
      return value;
    } else {
      return '$value';
    }
  }

  static final RegExp _regExpNotNumber = RegExp(r'[^0-9\.\+\-eENna]');

  /// Tries to parse an [int].
  /// - Returns [def] if [value] is invalid.
  static int? parseInt(Object? value, [int? def]) {
    if (value == null) return def;

    if (value is int) {
      return value;
    } else if (value is num) {
      return value.toInt();
    } else if (value is DateTime) {
      return value.millisecondsSinceEpoch;
    } else {
      var n = _parseNumString(value);
      return n?.toInt() ?? def;
    }
  }

  static num? _parseNumString(Object value) {
    var s = _valueAsString(value);
    if (s.isEmpty) {
      return null;
    }

    var n = num.tryParse(s);
    if (n == null) {
      s = s.replaceAll(_regExpNotNumber, '');
      n = num.tryParse(s);
    }

    return n;
  }

  static String _valueAsString(Object value) {
    String s;
    if (value is String) {
      s = value.trim();
    } else {
      s = '$value'.trim();
    }
    return s;
  }

  /// Tries to parse a [double].
  /// - Returns [def] if [value] is invalid.
  static double? parseDouble(Object? value, [double? def]) {
    if (value == null) return def;

    if (value is double) {
      return value;
    } else if (value is num) {
      return value.toDouble();
    } else if (value is DateTime) {
      return value.millisecondsSinceEpoch.toDouble();
    } else {
      var n = _parseNumString(value);
      return n?.toDouble() ?? def;
    }
  }

  /// Tries to parse a [num].
  /// - Returns [def] if [value] is invalid.
  static num? parseNum(Object? value, [num? def]) {
    if (value == null) return def;

    if (value is num) {
      return value;
    } else if (value is DateTime) {
      return value.millisecondsSinceEpoch;
    } else {
      var n = _parseNumString(value);
      return n ?? def;
    }
  }

  /// Tries to parse a [BigInt].
  /// - Returns [def] if [value] is invalid.
  static BigInt? parseBigInt(Object? value, [BigInt? def]) {
    if (value == null) return def;

    if (value is BigInt) {
      return value;
    } else if (value is num) {
      return BigInt.from(value);
    } else if (value is DateTime) {
      return BigInt.from(value.millisecondsSinceEpoch);
    } else {
      var s = _valueAsString(value);
      if (s.isEmpty) {
        return def;
      }

      var n = BigInt.tryParse(s);
      if (n == null) {
        s = s.replaceAll(_regExpNotNumber, '');
        n = BigInt.tryParse(s);
      }

      return n ?? def;
    }
  }

  /// Tries to parse a [bool].
  /// - Returns [def] if [value] is invalid.
  static bool? parseBool(Object? value, [bool? def]) {
    if (value == null) return def;

    if (value is bool) {
      return value;
    } else if (value is num) {
      return value > 0;
    } else {
      var s = _valueAsString(value).toLowerCase();
      if (s.isEmpty || s == 'null' || s == 'empty') return def;

      if (s == 'true' ||
          s == 't' ||
          s == 'yes' ||
          s == 'y' ||
          s == '1' ||
          s == '+' ||
          s == 'ok') {
        return true;
      }

      if (s == 'false' ||
          s == 'f' ||
          s == 'no' ||
          s == 'n' ||
          s == '0' ||
          s == '-1' ||
          s == '-' ||
          s == 'fail' ||
          s == 'error' ||
          s == 'err') {
        return false;
      }

      var n = _parseNumString(value);
      if (n != null) {
        return n > 0;
      }

      return def;
    }
  }

  /// Tries to parse a [DateTime].
  /// - Returns [def] if [value] is invalid.
  static DateTime? parseDateTime(Object? value, [DateTime? def]) {
    if (value == null) return def;

    if (value is DateTime) {
      return value;
    } else if (value is int) {
      return DateTime.fromMillisecondsSinceEpoch(value);
    } else {
      var s = '$value'.trim();
      var ms = int.tryParse(s);
      if (ms != null) {
        return DateTime.fromMillisecondsSinceEpoch(ms);
      }
      return DateTime.tryParse(s) ?? def;
    }
  }
}


lib/src/args_simple_io.dart
import 'dart:io';
import 'dart:typed_data';
import 'dart:convert' as dart_convert;

import 'package:yaml/yaml.dart';

import 'args_simple_base.dart';
import 'type_parser_io.dart';

extension ArgsSimpleExtension on ArgsSimple {
  File? argumentAsFile(int index, [File? def]) =>
      argument<File>(index, def, TypeParserIO.parseFile);

  String? argumentAsFileContentString(int index, [File? def]) {
    var file = argumentAsFile(index, def);
    return file?.readAsStringSync();
  }

  Uint8List? argumentAsFileContentBytes(int index, [File? def]) {
    var file = argumentAsFile(index, def);
    return file?.readAsBytesSync();
  }

  dynamic argumentAsFileContentJSON(int index, [File? def]) {
    var s = argumentAsFileContentString(index, def);
    return s != null ? dart_convert.json.decode(s) : null;
  }

  dynamic argumentAsFileContentYAML(int index, [File? def]) {
    var s = argumentAsFileContentString(index, def);
    return s != null ? loadYaml(s) : null;
  }

  Directory? argumentAsDirectory(int index, [Directory? def]) =>
      argument<Directory>(index, def, TypeParserIO.parseDirectory);

  File? optionAsFile(String key, [File? def]) =>
      option<File>(key, def, TypeParserIO.parseFile);

  String? optionAsFileContentString(String key, [File? def]) {
    var file = optionAsFile(key, def);
    return file?.readAsStringSync();
  }

  dynamic optionAsFileContentJSON(String key, [File? def]) {
    var s = optionAsFileContentString(key, def);
    return s != null ? dart_convert.json.decode(s) : null;
  }

  dynamic optionAsFileContentYAML(String key, [File? def]) {
    var s = optionAsFileContentString(key, def);
    return s != null ? loadYaml(s) : null;
  }

  Uint8List? optionAsFileContentBytes(String key, [File? def]) {
    var file = optionAsFile(key, def);
    return file?.readAsBytesSync();
  }

  Directory? optionAsDirectory(String key, [Directory? def]) =>
      option<Directory>(key, def, TypeParserIO.parseDirectory);

  File? propertyAsFile(String key, [File? def]) =>
      property<File>(key, def, TypeParserIO.parseFile);

  String? propertyAsFileContentString(String key, [File? def]) {
    var file = propertyAsFile(key, def);
    return file?.readAsStringSync();
  }

  dynamic propertyAsFileContentJSON(String key, [File? def]) {
    var s = propertyAsFileContentString(key, def);
    return s != null ? dart_convert.json.decode(s) : null;
  }

  dynamic propertyAsFileContentYAML(String key, [File? def]) {
    var s = propertyAsFileContentString(key, def);
    return s != null ? loadYaml(s) : null;
  }

  Uint8List? propertyAsFileContentBytes(String key, [File? def]) {
    var file = propertyAsFile(key, def);
    return file?.readAsBytesSync();
  }

  Directory? propertyAsDirectory(String key, [Directory? def]) =>
      property<Directory>(key, def, TypeParserIO.parseDirectory);

  Set<String> loadEnvironment() {
    var envs = Platform.environment;

    var loadedKeys = <String>{};

    for (var e in envs.entries) {
      var v = envs[e.key]!;
      var k = ArgsSimple.normalizeKey(e.key);

      if (!properties.containsKey(k)) {
        putProperty(k, v);
        loadedKeys.add(k);
      }
    }

    return loadedKeys;
  }
}


lib/src/type_parser_io.dart
import 'dart:io';

import 'type_parser.dart';

class TypeParserIO {
  static void register() {
    TypeParser.registerParserResolver(parserResolver);
  }

  static TypeElementParser? parserResolver({Type? t, Object? obj, Type? type}) {
    if (obj != null) {
      if (obj is File) {
        return parseFile;
      } else if (obj is Directory) {
        return parseDirectory;
      }
    }

    type ??= obj?.runtimeType ?? t;

    if (type == File) {
      return parseFile;
    } else if (type == Directory) {
      return parseDirectory;
    }

    return null;
  }

  /// Tries to parse a [File].
  /// - Returns [def] if [value] is invalid.
  static File? parseFile(Object? value, [File? def]) {
    if (value == null) return def;

    if (value is File) {
      return value;
    } else if (value is Directory) {
      return File(value.path);
    } else {
      var s = value.toString().trim();
      if (s.isEmpty) return def;

      return File(s);
    }
  }

  /// Tries to parse a [File].
  /// - Returns [def] if [value] is invalid.
  static Directory? parseDirectory(Object? value, [Directory? def]) {
    if (value == null) return def;

    if (value is Directory) {
      return value;
    } else if (value is File) {
      return Directory(value.path);
    } else {
      var s = value.toString().trim();
      if (s.isEmpty) return def;

      return Directory(s);
    }
  }
}


lib/src/args_simple_base.dart
import 'dart:collection';
import 'dart:convert' as dart_convert;

import 'package:args_simple/src/type_parser.dart';

class ArgsSimple {
  final List<String> _arguments = <String>[];

  List<String> get arguments => UnmodifiableListView<String>(_arguments);

  final Map<String, Object> _options = <String, Object>{};

  Map<String, Object> get options =>
      UnmodifiableMapView<String, Object>(_options);

  final Set<String> _flags = {};

  Set<String> get flags => UnmodifiableSetView<String>(_flags);

  final Map<String, Object> _properties = <String, Object>{};

  Map<String, Object> get properties =>
      UnmodifiableMapView<String, Object>(_properties);

  ArgsSimple();

  factory ArgsSimple.fromEncodedJson(String encodedJson) {
    var json = dart_convert.json.decode(encodedJson);
    return ArgsSimple.fromJson(json);
  }

  factory ArgsSimple.fromJson(Object json) {
    if (json is List<String>) {
      return ArgsSimple.parse(json);
    } else if (json is List) {
      return ArgsSimple.parse(json.map((e) => '$e').toList());
    }

    var args = ArgsSimple();

    if (json is Map) {
      args._options.addAll(
          json.map((key, value) => MapEntry(key.toString(), value ?? '')));
    } else {
      var s = '$json';
      args._arguments.add(s);
    }

    return args;
  }

  static final RegExp _propertyDelimiter = RegExp(r'[=:]');

  ArgsSimple.parse([List<String> args = const <String>[]]) {
    final length = args.length;

    for (var i = 0; i < length;) {
      var a = args[i];

      if (a.startsWith('--')) {
        var k = normalizeKey(a.substring(2));
        var v = i < length - 1 ? args[i + 1] : '';
        var prev = _options[k];
        if (prev != null) {
          if (prev is List) {
            prev.add(v);
          } else {
            _options[k] = [prev, v];
          }
        } else {
          _options[k] = v;
        }
        i += 2;
      } else if (a.startsWith('-P') || a.startsWith('-D')) {
        var idx = a.indexOf(_propertyDelimiter);
        if (idx > 0) {
          var k = normalizeKey(a.substring(2, idx));
          var v = a.substring(idx + 1);
          _properties[k] = v;
        } else {
          var k = normalizeKey(a.substring(2));
          _properties[k] = '';
        }
        ++i;
      } else if (a.startsWith('-')) {
        var s = a.substring(1);
        var n = num.tryParse(s);
        if (n != null) {
          _arguments.add(a);
        } else {
          var k = normalizeKey(s);
          _flags.add(k);
        }
        ++i;
      } else {
        _arguments.add(a);
        ++i;
      }
    }
  }

  bool get isEmpty =>
      _arguments.isEmpty &&
      _flags.isEmpty &&
      _options.isEmpty &&
      _properties.isEmpty;

  bool get isNotEmpty => !isEmpty;

  void addArgument(String value) => _arguments.add(value);

  T? argument<T>(int index, [T? def, TypeElementParser? parser]) {
    if (index >= _arguments.length) return def;
    var value = _arguments[index];

    parser ??= TypeParser.parserFor<T>(obj: def);
    if (parser != null) {
      return parser(value) as T;
    } else {
      return value as T;
    }
  }

  bool argumentMatches(int index, Pattern pattern) {
    var val = argumentAsString(index);
    if (val == null) return false;

    return _matches(pattern, val);
  }

  int? argumentAsInt(int index, [int? def]) =>
      argument<int>(index, def, TypeParser.parseInt);

  double? argumentAsDouble(int index, [double? def]) =>
      argument<double>(index, def, TypeParser.parseDouble);

  String? argumentAsString(int index, [String? def]) =>
      argument<String>(index, def, (s) => s);

  bool? argumentAsBool(int index, [bool? def]) =>
      argument<bool>(index, def, TypeParser.parseBool);

  List? argumentAsList(int index, [List? def]) =>
      argument<List>(index, def, TypeParser.parseList);

  Set? argumentAsSet(int index, [Set? def]) =>
      argument<Set>(index, def, TypeParser.parseSet);

  Map? argumentAsMap(int index, [Map? def]) =>
      argument<Map>(index, def, TypeParser.parseMap);

  void putOption(String key, Object value) =>
      _options[normalizeKey(key)] = value;

  T? option<T>(String key, [T? def, TypeElementParser? parser]) {
    var value = _options[key];

    if (value == null) {
      var keyLC = normalizeKey(key);
      value = _options[keyLC];
    }

    if (value == null) return def;

    parser ??= TypeParser.parserFor<T>(obj: def);
    if (parser != null) {
      return parser(value) as T;
    } else {
      return value as T;
    }
  }

  bool optionMatches(String key, Pattern pattern) {
    var val = optionAsString(key);
    if (val == null) return false;

    return _matches(pattern, val);
  }

  int? optionAsInt(String key, [int? def]) =>
      option<int>(key, def, TypeParser.parseInt);

  double? optionAsDouble(String key, [double? def]) =>
      option<double>(key, def, TypeParser.parseDouble);

  String? optionAsString(String key, [String? def]) =>
      option<String>(key, def, TypeParser.parseString);

  bool? optionAsBool(String key, [bool? def]) =>
      option<bool>(key, def, TypeParser.parseBool);

  List? optionAsList(String key, [List? def]) =>
      option<List>(key, def, TypeParser.parseList);

  Set? optionAsSet(String key, [Set? def]) =>
      option<Set>(key, def, TypeParser.parseSet);

  Map? optionAsMap(String key, [Map? def]) =>
      option<Map>(key, def, TypeParser.parseMap);

  void addFlag(String key) => _flags.add(normalizeKey(key));

  bool flag<T>(String key) {
    if (_flags.contains(key)) return true;
    var keyLC = normalizeKey(key);
    return _flags.contains(keyLC);
  }

  bool? flagOr<T>(String key, bool? def) {
    if (_flags.contains(key)) return true;
    var keyLC = normalizeKey(key);
    var found = _flags.contains(keyLC);
    if (found) return true;
    return def;
  }

  void putProperty(String key, Object value) =>
      _properties[normalizeKey(key)] = value;

  T? property<T>(String key, [T? def, TypeElementParser? parser]) {
    var value = _properties[key];

    if (value == null) {
      var keyLC = normalizeKey(key);
      value = _properties[keyLC];
    }

    if (value == null) return def;

    parser ??= TypeParser.parserFor<T>(obj: def);
    if (parser != null) {
      return parser(value) as T;
    } else {
      return value as T;
    }
  }

  bool propertyMatches(String key, Pattern pattern) {
    var val = propertyAsString(key);
    if (val == null) return false;

    return _matches(pattern, val);
  }

  int? propertyAsInt(String key, [int? def]) =>
      property<int>(key, def, TypeParser.parseInt);

  double? propertyAsDouble(String key, [double? def]) =>
      property<double>(key, def, TypeParser.parseDouble);

  String? propertyAsString(String key, [String? def]) =>
      property<String>(key, def, TypeParser.parseString);

  bool? propertyAsBool(String key, [bool? def]) =>
      property<bool>(key, def, TypeParser.parseBool);

  List? propertyAsList(String key, [List? def]) =>
      property<List>(key, def, TypeParser.parseList);

  Set? propertyAsSet(String key, [Set? def]) =>
      property<Set>(key, def, TypeParser.parseSet);

  Map? propertyAsMap(String key, [Map? def]) =>
      property<Map>(key, def, TypeParser.parseMap);

  bool _matches(Pattern pattern, String val) {
    if (pattern is RegExp) {
      return pattern.hasMatch(val);
    } else {
      return pattern.matchAsPrefix(val) != null;
    }
  }

  static final _regExpNonNormalizedKey = RegExp(r'[^a-z0-9]');

  static String normalizeKey(String key) =>
      key.toLowerCase().replaceAll(_regExpNonNormalizedKey, '').trim();

  List<String> toList() {
    return [
      ..._arguments,
      ..._flags.map((e) => '-$e'),
      ..._options.entries.expand((e) => ['--${e.key}', e.value.toString()]),
      ..._properties.entries.map((e) => '-P${e.key}=${e.value}'),
    ];
  }

  Map<String, Object> toJson() => {
        if (_arguments.isNotEmpty) 'arguments': [..._arguments],
        if (_flags.isNotEmpty) 'flags': {..._flags},
        if (_options.isNotEmpty) 'options': {..._options},
        if (_properties.isNotEmpty) 'properties': {..._properties},
      };

  @override
  String toString() => toList().toString();
}


example/args_simple_example.dart
import 'dart:io';

import 'package:args_simple/args_simple_io.dart';

/// Examples of valid values for the parameter `args`:
///
/// - `/path/config.json --port 81 -verbose`
/// - `/path/config.json -verbose --sys-config /my/sys-config-dir`
/// - `/path/config.json root`
void main(List<String> argsOrig) {
  var args = ArgsSimple.parse(argsOrig);

  if (args.isEmpty) {
    print(
        'USAGE: [%configFile.json] [%user] [--port] [--sys-config %systemConfig] [-verbose]');
    exit(0);
  }

  // Argument #0 is a JSON `File`:
  var config = args.argumentMatches(0, RegExp(r'.json$'))
      ? args.argumentAsFileContentJSON(0)!
      : {};

  // Argument #1 is an optional `String`:
  var user = args.argumentAsString(1, 'guest');

  // Option `sys-config` is a `Directory`:
  var systemConfigDir = args.optionAsDirectory(
      'sys-config', Directory('/default/sys-config-dir'))!;

  // Option `--port` is a `int`, with 8080 as default:
  var port = args.optionAsInt('port', 8080);

  // Check for flag `-verbose`:
  var verbose = args.flag('verbose');

  if (verbose) {
    print('-- Config: $config');
    print('-- User: $user');
    print('-- System Config Dir: $systemConfigDir');
    print('-- Port: $port');
    print('-- Verbose: $verbose');
    print(args);
  }
}
